File: app/src/main/java/io/github/ovso/righttoknow/ui/newsdetail/DetailNewsActivity.java
package io.github.ovso.righttoknow.ui.newsdetail;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.text.Html;
import android.text.SpannableString;
import android.view.MenuItem;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import androidx.annotation.Nullable;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import butterknife.BindView;
import io.github.ovso.righttoknow.R;
import io.github.ovso.righttoknow.framework.AdsActivity;
import io.github.ovso.righttoknow.ui.main.news.model.News;

public class DetailNewsActivity extends AdsActivity {
  @BindView(R.id.webview)
  WebView webView;

  @BindView(R.id.swipe_refresh)
  SwipeRefreshLayout swipeRefresh;

  private News news;

  @Override
  protected int getLayoutResId() {
    return R.layout.activity_detail_news;
  }

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    Intent intent = getIntent();
    if (intent.hasExtra("news")) {
      setInit();
      setWebView();
      setSwipeRefresh();
      setTitle();
      loadUrl();
    }
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    activityFinish();
    return super.onOptionsItemSelected(item);
  }

  private void activityFinish() {
    finish();
    webView.stopLoading();
  }

  private void setTitle() {
    if (getSupportActionBar() != null) {
      getSupportActionBar().setDisplayHomeAsUpEnabled(true);
      getSupportActionBar().setTitle(new SpannableString(Html.fromHtml(news.getTitle())));
    }
  }

  private void loadUrl() {
    swipeRefresh.setRefreshing(true);
    webView.loadUrl(news.getLink());
  }

  private void setInit() {
    news = (News) getIntent().getSerializableExtra("news");
  }

  private void setSwipeRefresh() {
    swipeRefresh.setOnRefreshListener(() -> webView.loadUrl(news.getLink()));
    swipeRefresh.setColorSchemeResources(R.color.colorPrimary);
  }

  @SuppressLint("SetJavaScriptEnabled")
  private void setWebView() {
    WebSettings settings = webView.getSettings();
    settings.setJavaScriptEnabled(true);
    webView.setWebChromeClient(new MyWebChromeClient(webClientListener));
    webView.setWebViewClient(new MyWebViewClient(webClientListener));
  }

  static class MyWebChromeClient extends WebChromeClient {

    private final WebClientListener l;

    public MyWebChromeClient(WebClientListener l) {
      this.l = l;
    }

    @Override
    public void onProgressChanged(WebView view, int newProgress) {
      super.onProgressChanged(view, newProgress);
      if (l != null) l.onProgressChanged(newProgress);
    }
  }

  static class MyWebViewClient extends WebViewClient {

    private final WebClientListener l;

    MyWebViewClient(WebClientListener l) {
      this.l = l;
    }

    @Override
    public void onPageStarted(WebView view, String url, Bitmap favicon) {
      super.onPageStarted(view, url, favicon);
      if (l != null) l.onPageStarted();
    }

    @Override
    public void onPageFinished(WebView view, String url) {
      super.onPageFinished(view, url);
      if (l != null) l.onPageFinish();
    }
  }

  private SimpleWebClientListener webClientListener =
      new SimpleWebClientListener() {
        @Override
        public void onPageStarted() {
          super.onPageStarted();
        }

        @Override
        public void onPageFinish() {
          super.onPageFinish();
        }

        @Override
        public void onProgressChanged(int newProgress) {
          super.onProgressChanged(newProgress);
        }
      };

  static class SimpleWebClientListener implements WebClientListener {

    @Override
    public void onPageStarted() {}

    @Override
    public void onPageFinish() {}

    @Override
    public void onProgressChanged(int newProgress) {}
  }

  interface WebClientListener {
    void onPageStarted();

    void onPageFinish();

    void onProgressChanged(int newProgress);
  }
}

File: app/src/main/java/io/github/ovso/righttoknow/ui/newsdetail/DetailNewsActivity.kt
package io.github.ovso.righttoknow.ui.newsdetail

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.os.Bundle
import android.text.SpannableString
import android.view.MenuItem
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.core.text.HtmlCompat
import io.github.ovso.righttoknow.R
import io.github.ovso.righttoknow.framework.AdsActivity
import io.github.ovso.righttoknow.ui.main.news.model.News
import kotlinx.android.synthetic.main.app_bar_detail_news.*
import kotlinx.android.synthetic.main.content_news_detail.*

class DetailNewsActivity : AdsActivity() {

  private lateinit var news: News

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_detail_news)
    setSupportActionBar(toolbar)
    pb_new_detail.hide()
    showAds()
    val intent = intent
    if (intent.hasExtra("news")) {
      setInit()
      setWebView()
      setTitle()
      loadUrl()
    }
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    activityFinish()
    return super.onOptionsItemSelected(item)
  }

  private fun activityFinish() {
    finish()
    webview.stopLoading()
  }

  private fun setTitle() {
    supportActionBar?.run {
      setDisplayHomeAsUpEnabled(true)
      title = SpannableString(
        HtmlCompat.fromHtml(news.title, HtmlCompat.FROM_HTML_MODE_COMPACT)
      )
    }
  }

  private fun loadUrl() {
    webview.loadUrl(news.link)
  }

  private fun setInit() {
    news = intent.getSerializableExtra("news") as News
  }

  @SuppressLint("SetJavaScriptEnabled")
  private fun setWebView() {
    val settings = webview.settings
    settings.javaScriptEnabled = true
    webview.webChromeClient = MyWebChromeClient(webClientListener)
    webview.webViewClient = MyWebViewClient(webClientListener)
  }

  override fun onDestroy() {
    webview.destroy()
    super.onDestroy()
  }

  internal class MyWebChromeClient(private val l: WebClientListener?) : WebChromeClient() {
    override fun onProgressChanged(view: WebView, newProgress: Int) {
      super.onProgressChanged(view, newProgress)
      l?.onProgressChanged(newProgress)
    }

  }

  internal class MyWebViewClient(private val l: WebClientListener?) : WebViewClient() {
    override fun onPageStarted(
      view: WebView?,
      url: String?,
      favicon: Bitmap?
    ) {
      super.onPageStarted(view, url, favicon)
      l?.onPageStarted()
    }

    override fun onPageFinished(view: WebView, url: String) {
      super.onPageFinished(view, url)
      l?.onPageFinish()
    }

  }

  private val webClientListener: SimpleWebClientListener = object : SimpleWebClientListener() {
    override fun onPageStarted() {
      super.onPageStarted()
      pb_new_detail.show()
    }

    override fun onPageFinish() {
      super.onPageFinish()
      pb_new_detail.hide()
    }

    override fun onProgressChanged(newProgress: Int) {
      super.onProgressChanged(newProgress)
      pb_new_detail.progress = newProgress
    }
  }

  internal open class SimpleWebClientListener : WebClientListener {
    override fun onPageStarted() {}
    override fun onPageFinish() {}
    override fun onProgressChanged(newProgress: Int) {}
  }

  internal interface WebClientListener {
    fun onPageStarted()
    fun onPageFinish()
    fun onProgressChanged(newProgress: Int)
  }

  override fun onBackPressed() {
    if (!webview.canGoBack()) {
      super.onBackPressed()
    } else {
      webview.goBack()
    }
  }
}
